File: src/main/java/uk/laxd/dndSimulator/character/CharacterConfigBuilder.java
package uk.laxd.dndSimulator.character;

import uk.laxd.dndSimulator.ability.Ability;
import uk.laxd.dndSimulator.equipment.Weapon;
import uk.laxd.dndSimulator.feature.barbarian.Rage;
import uk.laxd.dndSimulator.feature.barbarian.RecklessAttack;
import uk.laxd.dndSimulator.feature.barbarian.UnarmoredDefence;

public class CharacterConfigBuilder {

    private CharacterConfig characterConfig;

    private CharacterConfigBuilder(){}

    public static CharacterConfigBuilder newCharacter(String name) {
        CharacterConfigBuilder builder = new CharacterConfigBuilder();
        builder.characterConfig = new CharacterConfig();
        builder.characterConfig.setName(name);

        // Set all abilities to 10 by default
        builder.characterConfig.setAbilityScore(Ability.STRENGTH, 10);
        builder.characterConfig.setAbilityScore(Ability.DEXTERITY, 10);
        builder.characterConfig.setAbilityScore(Ability.CONSTITUTION, 10);
        builder.characterConfig.setAbilityScore(Ability.INTELLIGENCE, 10);
        builder.characterConfig.setAbilityScore(Ability.WISDOM, 10);
        builder.characterConfig.setAbilityScore(Ability.CHARISMA, 10);

        return builder;
    }

    public CharacterConfigBuilder withLevels(int level, CharacterClass characterClass) {
        this.characterConfig.addLevel(level, characterClass);
        return this;
    }

    public CharacterConfigBuilder withArmourClass(short ac) {
        characterConfig.setArmourClass(ac);
        return this;
    }

    public CharacterConfigBuilder withAbilityScores(int str, int dex, int con, int intelligence, int wis, int cha) {
        characterConfig.setAbilityScore(Ability.STRENGTH, str);
        characterConfig.setAbilityScore(Ability.DEXTERITY, dex);
        characterConfig.setAbilityScore(Ability.CONSTITUTION, con);
        characterConfig.setAbilityScore(Ability.INTELLIGENCE, intelligence);
        characterConfig.setAbilityScore(Ability.WISDOM, wis);
        characterConfig.setAbilityScore(Ability.CHARISMA, cha);

        // Set default HP

        return this;
    }

    public CharacterConfigBuilder withAbilityScore(Ability ability, int score) {
        characterConfig.setAbilityScore(ability, score);
        return this;
    }

    public CharacterConfigBuilder withHp(short hp) {
        characterConfig.setHp(hp);
        return this;
    }

    // TODO: Change to weapon lookups by ID/Name? Custom weapon definitions?
    public CharacterConfigBuilder withWeapon(Weapon weapon) {
        //characterConfig.setWeapon(weapon);
        return this;
    }

    public CharacterConfig build() {
        return characterConfig;
    }

    // TODO: Move to Character creation
//    public Character build() {
//        // Set default HP if not already set
//        if(character.getMaxHp() == 0) {
//            int hp = character.getHitDie().getMaxValue()
//                    + (character.getLevel() * character.getAbilityModifier(Ability.CONSTITUTION))
//                    + ((character.getLevel() - 1) * ((character.getHitDie().getMaxValue()/2)+1));
//
//            character.setMaxHp(hp);
//        }
//
//        character.setHp(character.getMaxHp());
//
//        // Tell all features they have been set
//        character.getFeatures()
//                .forEach(f -> f.onCreate(character));
//
//        return character;
//    }

}

File: app/src/main/kotlin/uk/laxd/dndSimulator/character/CharacterConfigBuilder.kt
package uk.laxd.dndSimulator.character

import uk.laxd.dndSimulator.ability.Ability
import uk.laxd.dndSimulator.equipment.Weapon

class CharacterConfigBuilder private constructor() {

    private var characterConfig: CharacterConfig = CharacterConfig()

    fun withLevels(level: Int, characterClass: CharacterClass): CharacterConfigBuilder {
        characterConfig.addLevel(level, characterClass)
        return this
    }

    fun withArmourClass(ac: Short): CharacterConfigBuilder {
        characterConfig.armourClass = ac
        return this
    }

    fun withAbilityScores(str: Int, dex: Int, con: Int, intelligence: Int, wis: Int, cha: Int): CharacterConfigBuilder {
        characterConfig.abilityScores[Ability.STRENGTH] = str
        characterConfig.abilityScores[Ability.DEXTERITY] = dex
        characterConfig.abilityScores[Ability.CONSTITUTION] = con
        characterConfig.abilityScores[Ability.INTELLIGENCE] = intelligence
        characterConfig.abilityScores[Ability.WISDOM] = wis
        characterConfig.abilityScores[Ability.CHARISMA] = cha
        return this
    }

    fun withAbilityScore(ability: Ability, score: Int): CharacterConfigBuilder {
        characterConfig.abilityScores[ability] = score
        return this
    }

    fun withHp(hp: Short): CharacterConfigBuilder {
        characterConfig.hp = hp
        return this
    }

    // TODO: Change to weapon lookups by ID/Name? Custom weapon definitions?
    fun withWeapon(weapon: Weapon): CharacterConfigBuilder {
        characterConfig.weapon = weapon;
        return this
    }

    fun build(): CharacterConfig {
        return characterConfig
    }

    companion object {
        fun newCharacter(name: String): CharacterConfigBuilder {
            val builder = CharacterConfigBuilder()

            builder.characterConfig.apply {
                this.name = name

                // Set all abilities to 10 by default
                abilityScores[Ability.STRENGTH] = 10
                abilityScores[Ability.DEXTERITY] = 10
                abilityScores[Ability.CONSTITUTION] = 10
                abilityScores[Ability.INTELLIGENCE] = 10
                abilityScores[Ability.WISDOM] = 10
                abilityScores[Ability.CHARISMA] = 10
            }

            return builder
        }
    }
}